powerstore/datasource_volume.go

```diff
@@ -0,0 +1,505 @@
+									},
+								},
+							},
+						},
+						"volume_groups": schema.ListNestedAttribute{
+							Description:         "Specifies the VolumeGroup for a volume.",
+							MarkdownDescription: "Specifies the VolumeGroup for a volume.",
+							Computed:            true,
+							NestedObject: schema.NestedAttributeObject{
+								Attributes: map[string]schema.Attribute{
+									"id": schema.StringAttribute{
+										Description:         "Unique identifier of the volume group.",
+										MarkdownDescription: "Unique identifier of the volume group.",
+										Computed:            true,
+									},
+									"name": schema.StringAttribute{
+										Description:         "Name of the volume group.",
+										MarkdownDescription: "Name of the volume group.",
+										Computed:            true,
+									},
+									"description": schema.StringAttribute{
+										Description:         "Description for the volume group.",
+										MarkdownDescription: "Description for the volume group.",
+										Computed:            true,
+									},
+								},
+							},
+						},
+						"protection_policy": schema.ObjectAttribute{
+							Description:         "Specifies the protection policy associated for a volume.",
+							MarkdownDescription: "Specifies the protection policy associated for a volume.",
+							Computed:            true,
+							AttributeTypes: map[string]attr.Type{
+								"id":          types.StringType,
+								"name":        types.StringType,
+								"description": types.StringType,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (d *volumeDataSource) Configure(_ context.Context, req datasource.ConfigureRequest, _ *datasource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	d.client = req.ProviderData.(*client.Client)
+}
+
+func (d *volumeDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var state volumeDataSourceModel
+	var volumes []gopowerstore.Volume
+	var volume gopowerstore.Volume
+	var err error
+
+	diags := req.Config.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	//Read the volumes based on volume id/name and if nothing is mentioned, then it returns all the volumes
+	if state.Name.ValueString() != "" {
+		volume, err = d.client.PStoreClient.GetVolumeByName(context.Background(), state.Name.ValueString())
+	} else if state.ID.ValueString() != "" {
+		volume, err = d.client.PStoreClient.GetVolume(context.Background(), state.ID.ValueString())
+	} else {
+		volumes, err = d.client.PStoreClient.GetVolumes(context.Background())
+	}
+	//check if there is any error while getting the volume
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to Read PowerStore Volumes",
+			err.Error(),
+		)
+		return
+	}
+	volumes = append(volumes, volume)
+	if len(volumes) == 0 {
+		resp.Diagnostics.AddError(
+			"No Volume found corresponding to the given name/ID",
```

Comment:
gopowerstore is not throwing an error if volume did not exist with given id/name?